File: pset6/credit.py
from cs50 import get_int
from sys import exit


def get_positive_int():
    while True: 
        card_number = get_int("Number: ")
        if card_number > 0:
            return card_number


def main():
    iter = 0
    sum_odd = 0
    sum_even = 0
    d1 = 0
    d2 = 0
    doubled_digit = 0
    last_digit = 0
    sum_digit = 0
    check_sum = 0
    card_number = get_positive_int()
   
    while card_number > 0:
        digit = card_number % 10
        
        if iter % 2 == 0:
            sum_odd += digit
        else: 
            doubled_digit = digit * 2
            if doubled_digit >= 10:
                last_digit = doubled_digit % 10
                sum_digit = 1 + last_digit
                sum_even += sum_digit
            else:
                sum_even += doubled_digit
        
        card_number //= 10
        d2 = d1
        d1 = digit
        iter += 1 
        
    check_sum = sum_odd + sum_even
    if check_sum % 10 != 0:
        print("INVALID")
        exit(0)
   
    if iter == 15 and d1 == 3 and (d2 == 4 or d2 == 7):
        print("AMEX")
    elif iter == 16 and d1 == 5 and d2 >= 1 and d2 <= 5:
        print("MASTERCARD")
    elif (iter == 13 or iter == 16) and d1 == 4:
        print("VISA")
    else:
        print("INVALID")


main()

File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    long card_number;
    do
    {
        card_number = get_long("Number: ");
    }
    while (card_number < 0);
    
    int iter = 0; 
    int sum_odd = 0;
    int sum_even = 0;
    int d1 = 0;
    int d2 = 0;

    while (card_number > 0)
    {
        long digit = card_number % 10;
        if (iter % 2 == 0)
        {
            sum_odd += digit; 
        }
        else 
        {
            int doubled_digit = digit * 2;
            if (doubled_digit >= 10)
            {
                int last_digit = doubled_digit % 10;
                int sum_digit = 1 + last_digit;
                sum_even += sum_digit;
            }
            else
            {
                sum_even += doubled_digit;
            }
        }
        card_number /= 10;
        d2 = d1;
        d1 = digit;
        iter ++ ;
        
    }
     
    int check_sum = sum_odd + sum_even;
    if (check_sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
   
   
    if (iter == 15 && d1 == 3 && (d2 == 4 || d2 == 7))
    {
        printf("AMEX\n");
    }
    else if (iter == 16 && d1 == 5 && d2 >= 1 && d2 <= 5)
    {
        printf("MASTERCARD\n");
    }
    else if ((iter == 13 || iter == 16) && d1 == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
